### ISSUE TITLE
SemVer for elasticsearch-dsl

### ISSUE THREAD

**Issue by 3lnc** [elastic/elasticsearch-dsl-py#931]
With `6.2.0` release I've suddenly realised that dsl in not following semver (minor version introduced backward-incompatible changes).

I thought that semver is accepted for all products under elastic umbrella, at least starting from "synchronized" releases (when all ELK stack rolls out with same versions). Isn't this a thing? Or dsl is not supported by this level of effort — as not an official client?

**Comment by msholty-fd** [comment]
The documentation recommends that you install with this in your requirements.txt:
```
# Elasticsearch 6.x
elasticsearch-dsl>=6.0.0,<7.0.0
```

Yet the latest release is not backwards compatible, breaking anyone doing a fresh pip install on each release. There are probably going to be a lot of people waking up this morning to things broken and not knowing why, like my company did this morning.

In the end, the latest release should have been a major version bump, or the documentation needs to change to recommend installing with a specific version number to avoid this issue.

The issue I ran into this morning was that `DocType` was renamed to `Document`, thus giving us this error:

```
 File "/opt/atlassian/pipelines/agent/build/app/databases/elasticsearch/models/foo.py", line 4, in <module>
   from elasticsearch_dsl import DocType, analyzer, tokenizer
ImportError: cannot import name DocType
```

**Comment by 3lnc** [comment]
@msholty-fd I recommend to pin you requirements to `>=6,<6.2`, all you've wrote is true.

Important thing here — even `6.2.1` release that probably will be available later today will not be backward-compatible.

**Comment by HonzaKral** [comment]
I understand, and share, the frustration. Unfortunately we made the decision to keep the major version compatible with the version of elasticsearch and that limits what we can do.

Only thing we could do is we could move to a 4-digit version in the future, where first is ES version, then major, minor, and bugfix. Would that help?

**Comment by msholty-fd** [comment]
@HonzaKral I think having 4 digits on your version isn't optimal. I would recommend just being very clear in the documentation what the versioning means. Most people assume that if you have versioning like `x.y.z`, it's SemVer, and are confused when it's actually not.

Why does the documentation recommend installing such a broad range, like `elasticsearch-dsl>=6.0.0,<7.0.0`? Maybe that is a good start, but once you `pip install elasticsearch-dsl>=6.0.0,<7.0.0`, you should check to see which version was actually installed and then lock it down by putting an actual version number in your `requirements.txt` like `elasticsearch-dsl==6.2.0`

**Comment by 3lnc** [comment]
@msholty-fd I;m not sure if you're talking about documentation for es-dsl or general approach, but just in case: pinning dependencies for _apps_ is common and good practice. You can never guarantee that 4rd-party libs will follow semver. Sometimes by mistake, sometimes willingly, sometimes simply by not knowing of such.

Important things: even with pinned versions, you should consider using hash-based checks (pip [supports this](https://pip.pypa.io/en/stable/reference/pip_install/#hash-checking-mode) starting from 8.0). Because there's possibility to get different code labeled with same version in PyPI. Again reasons is different.

And there's some tools to pin/updated pinned versions semi-automatically.

**Comment by msholty-fd** [comment]
@3lnc I was mentioning this repo's main README.md. It says to install this library with `elasticsearch-dsl>=6.0.0,<7.0.0`, which is where my issue yesterday came from originally. When the new version released, we got that new version unexpectedly in our pipeline and even though we have unit tests, the timing seemed to line up perfectly where this new version snuck into production and a lot of things came crashing down as a result.

We are going to soon migrate to using Docker for this application, which will alleviate the above issue, but in the meantime, we will look into hash-based checks. Thanks for the suggestion.

**Comment by Paulius-Maruska** [comment]
Or you could use Pipenv. I don't actually use Pipenv in production, but all of our developers use Pipenv locally. In our Pipfile we have `elasticsearch-dsl = "<7.0.0,>=6.0.0"` as recommended here in the README file. But when deploying - we are installing dependencies with `pip install -r requirements.txt`.
We only ever update `requirements.txt` file with `pipenv run pip freeze > requirements.txt` command locally, when we know that everything works.

Therefor, all our requirements are pinned. Even when we play around with different versions in our Pipfile - as long as `requirements.txt` stays untouched - our production environment will not (or should not) break.

**Comment by HonzaKral** [comment]
Thank you all for feedback, it makes perfect sense and I have to apologize again for the incompatibilities `6.3.0` brought in a minor version.

While there are technical tools to pin dependencies (and I do highly recommend that for production) that is no excuse to break people's code. I didn't make the decision lightly but I felt it was warranted given the amount of changes and fixes that became available in `6.3`.  Going forwards I will be even more cautious in introducing breaking changes outside of major version.

If I ever decide that it is required to introduce backwards incompatible changes (outside of severe bug fixes and possible security patches) I will consider just breaking the elasticsearch version sync instead.
